File: business/src/main/java/com/pij/lottieshow/model/LottieFile.java
package com.pij.lottieshow.model;


import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;

import java.io.File;
import java.net.URI;

import rx.Single;

import static rx.Single.just;

@AutoValue
public abstract class LottieFile {

    private Single<String> content;

    public static LottieFile create(LottieFile source, String content) {
        return create(source, just(content));
    }

    public static LottieFile create(LottieFile source, @NonNull Single<String> content) {
        return create(source.id(), source.label(), content);
    }

    public static LottieFile create(File id) {
        return create(id.toURI());
    }

    public static LottieFile create(URI id) {
        return create(id, null);
    }

    public static LottieFile create(URI id, String label) {
        return create(id, label, (String)null);
    }

    public static LottieFile create(URI id, String label, String content) {
        return create(id, label, just(content));
    }

    public static LottieFile create(URI id, String label, @NonNull Single<String> content) {
        LottieFile result = new AutoValue_LottieFile(id, label);
        result.content = content;
        return result;
    }

    public abstract URI id();

    @Nullable
    public abstract String label();

    @NonNull
    public final Single<String> content() {
        return content;
    }
}

File: business/src/main/java/com/pij/lottieshow/model/LottieFile.kt
package com.pij.lottieshow.model


import rx.Single
import rx.Single.just
import java.io.File
import java.net.URI

fun URI.toLottie(): LottieFile {
    return LottieFile(this)
}

data class LottieFile(val id: URI, val label: String? = null, val content: Single<String>? = null) {

    companion object {

        fun create(source: LottieFile, content: String): LottieFile = create(source, just(content))

        fun create(source: LottieFile, content: Single<String>): LottieFile = LottieFile(source.id, source.label,
                                                                                         content)

        fun create(id: File): LottieFile {
            return id.toURI().toLottie()
        }

        fun create(id: URI): LottieFile {
            return id.toLottie()
        }

        //        @JvmOverloads fun create(id: URI, label: String? = null, content: String? = null): LottieFile {
        //            return LottieFile(id, label, just(content))
        //        }
        //
        fun create(id: URI, label: String, content: Single<String>): LottieFile {
            return LottieFile(id, label, content)
        }
    }
}
